### ISSUE TITLE
[Device Client] Disconnects and never reconnects

### ISSUE THREAD

**Issue by pizerg** [Azure/azure-iot-sdk-java#304]
- **OS and version used:** Windows 10, Raspbian

- **Java runtime used:** Java 1.8

- **SDK version used:** Device Client 1.13.2

# Description of the issue:
Tried with the following protocols (same or equivalent results with all of them): AMQPS_WS, MQTT_WS and MQTT

After a successful connection it runs for a while (less than 2 hours) and without a visible reason the client disconnects (see exception below), status callback is called with DISCONNECTED_RETRYING but it never reconnects again (internet connection never drops during the tests).

Besides that, after updating to 1.13.2 the log has been flooded with the following (even after the connection drops which shouldn't be saying "Successfully send all queued messages" as new messages are queued but they're never sent as the client has disconnected)

```
2018-08-13 12:19:13,226 IotHubSendTask [azure-iot-sdk-IotHubSendTask] TRACE: Now sending all queued messages to IoT Hub
2018-08-13 12:19:13,226 IotHubSendTask [azure-iot-sdk-IotHubSendTask] TRACE: Now invoking all queued callbacks
2018-08-13 12:19:13,226 IotHubSendTask [azure-iot-sdk-IotHubSendTask] TRACE: Successfully send all queued messages to IoT Hub
```

# Console log of the issue:
```
com.microsoft.azure.sdk.iot.device.exceptions.ProtocolException: Mqtt connection lost
	at com.microsoft.azure.sdk.iot.device.transport.mqtt.exceptions.PahoExceptionTranslator.convertToMqttException(PahoExceptionTranslator.java:75)
	at com.microsoft.azure.sdk.iot.device.transport.mqtt.Mqtt.connectionLost(Mqtt.java:346)
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.connectionLost(CommsCallback.java:292)
	at org.eclipse.paho.client.mqttv3.internal.ClientComms.shutdownConnection(ClientComms.java:423)
	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:181)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: Connection lost (32109) - java.io.EOFException
	... 8 more
Caused by: java.io.EOFException
	at java.io.DataInputStream.readByte(DataInputStream.java:267)
	at org.eclipse.paho.client.mqttv3.internal.wire.MqttInputStream.readMqttWireMessage(MqttInputStream.java:92)
	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:133)
	... 7 more
```

**Comment by timtay-microsoft** [comment]
We'll take a look into this soon, thanks for filing the issue.

**Comment by pizerg** [comment]
I've done further testing and I can add the following information to the issue, hope it's useful:

I started the same program (AMQPS_WS protocol) at the same time in different networks (in different cities as well), all of them started working normally, after a while all of them stopped sending/receiving messages at more or less the same timestamp (some seconds to a minute later). 

In all cases the iothub connection never recovered and internet connection never dropped.

Maybe it's possible the error relies on the Azure Iot Hub server side and not on the device client, if you need additional information (instance ids, etc) please provide a contact email.

These are some exceptions I found in the logs of the program (not in any particular order):

```
2018-08-14 12:11:31,175 AzureIoTHub [pool-109-thread-1] DEBUG: IoTHub connection state callback, status: DISCONNECTED_RETRYING reason: NO_NETWORK exception:
com.microsoft.azure.sdk.iot.device.exceptions.TransportException: Unknown transport exception occurred
	at com.microsoft.azure.sdk.iot.device.transport.amqps.AmqpsIotHubConnection.onLinkRemoteClose(AmqpsIotHubConnection.java:713)
	at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:176)
	at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
	at com.microsoft.azure.sdk.iot.device.transport.amqps.IotHubReactor.run(IotHubReactor.java:28)
	at com.microsoft.azure.sdk.iot.device.transport.amqps.AmqpsIotHubConnection$ReactorRunner.call(AmqpsIotHubConnection.java:808)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

```
2018-08-14 12:21:22,465 AzureIoTHub [pool-35-thread-1] DEBUG: Exception while sending IoTHub message
java.io.IOException: Could not open the connection
	at com.microsoft.azure.sdk.iot.device.DeviceIO.open(DeviceIO.java:164)
	at com.microsoft.azure.sdk.iot.device.InternalClient.open(InternalClient.java:130)
	at com.microsoft.azure.sdk.iot.device.DeviceClient.open(DeviceClient.java:307)
        ...
	at current.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.microsoft.azure.sdk.iot.device.exceptions.TransportException: Unknown transport exception occurred
	at com.microsoft.azure.sdk.iot.device.transport.amqps.AmqpsIotHubConnection.onLinkRemoteClose(AmqpsIotHubConnection.java:713)
	at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:176)
	at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
	at com.microsoft.azure.sdk.iot.device.transport.amqps.IotHubReactor.run(IotHubReactor.java:28)
	at com.microsoft.azure.sdk.iot.device.transport.amqps.AmqpsIotHubConnection$ReactorRunner.call(AmqpsIotHubConnection.java:808)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	... 3 more
```

```
2018-08-14 11:04:37,306 AmqpsIotHubConnection [azure-iot-sdk-ReactorRunner] DEBUG: Entered in method onDelivery
2018-08-14 11:04:37,385 AmqpsIotHubConnection [azure-iot-sdk-ReactorRunner] INFO : Reading the delivery event in Sender link, method name is onDelivery 
2018-08-14 11:04:37,385 AmqpsIotHubConnection [azure-iot-sdk-ReactorRunner] INFO : Is state of remote Delivery COMPLETE ? CONNECTED, method name is onDelivery 
2018-08-14 11:04:37,385 AmqpsIotHubConnection [azure-iot-sdk-ReactorRunner] INFO : Inform listener that a message has been sent to IoT Hub along with remote state, method name is onDelivery 
2018-08-14 11:04:37,388 IotHubTransport [azure-iot-sdk-ReactorRunner] ERROR: Exception encountered while receiving messages from service, method name is onMessageReceived
2018-08-14 11:04:37,390 IotHubTransport [azure-iot-sdk-ReactorRunner] ERROR: com.microsoft.azure.sdk.iot.device.exceptions.TransportException: Received response from service about a message that this client did not send
```

**Comment by timtay-microsoft** [comment]
What are your device clients doing during these long periods of time? Are they actively sending messages? If so, how frequently? What kind of authentication are you using (sas token, device key, or x509)?

**Comment by timtay-microsoft** [comment]
For the MQTT logs you provided, the SDK should have kept retrying to connect after encountering that particular exception. Was this really the last exception given to you by the SDK during the DISCONNECTED callback?

**Comment by pizerg** [comment]
@timtay-microsoft The clients are sending at least 1 message every 4 minutes and they're authenticated with device key.

As for MQTT that's the last exception that appears, it seems there's some serious problem with the reconnect, right now the only way to make iot hub work continuously and reconnect on connection loss is to treat DISCONNECTED_RETRYING as DISCONNECTED, closing the client en opening a new one.

**Comment by timtay-microsoft** [comment]
This may be due to an issue on the IotHub side as we have customers using the C# sdk facing a similar issue. We will need more time to resolve this issue, though.

**Comment by timtay-microsoft** [comment]
The default expiry time for the sas token that is automatically generated for you from a device key is 60 minutes, so I bet the SDK is having some trouble handling when the iot hub flags the connection as unauthorized (which happens within 5-10 minutes after the sas token actually expires).

@cesarcarc could you provide the full log you get during this scenario including all connection status updates you get. I need to see the exceptions that the current retry logic is encountering.

**Comment by timtay-microsoft** [comment]
I found the cause of the bug here, and it was with this SDK's reconnection logic and it did impact every protocol equally. I'll have this bug fix checked into master in time for the next release. Thanks for all of your notes on this issue, they helped a lot!

**Comment by cesarcarc** [comment]
@timtay-microsoft, I didn't post my logs because it was equal to what @pizerg had. 
Nice that you found the cause.

**Comment by rkanubaddhi** [comment]
@timtay-microsoft - when can we expect the next release? do you have a date?

**Comment by timtay-microsoft** [comment]
We plan on doing our next release on 8/30. Thanks for your patience on this issue.

**Comment by timtay-microsoft** [comment]
The fix for this issue has been checked into master.

**Comment by timtay-microsoft** [comment]
The fix for this issue has been released as of Device Client 1.14.0, so I'm closing this issue now.

**Comment by az-iot-builder-01** [comment]
@pizerg, @cesarcarc, @rkanubaddhi, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
